N/nativefiledialog_extended/build_tarballs.jl

```diff
@@ -0,0 +1,95 @@
+# Note that this script can accept some limited command-line arguments, run
+# `julia build_tarballs.jl --help` to see a usage message.
+using BinaryBuilder, Pkg
+
+name = "nativefiledialog_extended"
+version = v"1.1.1"
+
+# Collection of sources required to complete build
+sources = [
+    GitSource("https://github.com/btzy/nativefiledialog-extended.git", "5786fabceeaee4d892f3c7a16b243796244cdddc")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd ${WORKSPACE}/srcdir/nativefiledialog-extended
+# NATIVE   mingw, apple
+# PORTAL   linux ~armv6l
+# GTK3     linux ~armv6l, freebsd
+# PORTAL?  linux  armv6l
+
+function build_nfd()
+{
+    BUILD=${1}
+    PREFIX=${2}
+    NFD_PORTAL=${3}
+    NFD_APPEND_EXTENSION=${4}
+    cmake -B ${BUILD} \
+        -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+        -DCMAKE_BUILD_TYPE=Release \
+        -DCMAKE_CXX_STANDARD=11 \
+        -DCMAKE_C_FLAGS="-Wall -Wextra -pedantic" \
+        -DCMAKE_CXX_FLAGS="-Wall -Wextra -pedantic" \
+        -DBUILD_SHARED_LIBS=ON \
+        -DCMAKE_INSTALL_PREFIX=${PREFIX} \
+        -DNFD_PORTAL=${NFD_PORTAL} \
+        -DNFD_APPEND_EXTENSION=${NFD_APPEND_EXTENSION} \
+        .
+    cmake --build ${BUILD} --parallel ${nproc}
+    cmake --install ${BUILD}
+}
+
+# NATIVE and PORTAL
+BUILD="build"
+PREFIX="${prefix}"
+NFD_PORTAL=OFF
+NFD_APPEND_EXTENSION=OFF
+if [[ "${target}" == *linux* || "${target}" == *freebsd* ]] ; then
+    NFD_PORTAL=ON
+fi
+if [[ "${target}" != *armv6l* ]] ; then
+    build_nfd ${BUILD} ${PREFIX} ${NFD_PORTAL} ${NFD_APPEND_EXTENSION}
+fi
+
+# GTK3
+BUILD="build_gtk"
+if [[ "${target}" == *linux* ]] ; then
+    PREFIX="${prefix}/gtk"
+    mkdir ${PREFIX}
+fi
+NFD_PORTAL=OFF
+NFD_APPEND_EXTENSION=ON
+if [[ "${target}" != *armv6l* && "${target}" != *x86_64-*-musl* ]] ; then
+    build_nfd ${BUILD} ${PREFIX} ${NFD_PORTAL} ${NFD_APPEND_EXTENSION}
+    if [[ "${target}" == *mingw32* ]] ; then
+        install ${PREFIX}/bin/libnfd.${dlext} ${libdir}/libnfd_gtk.${dlext}
+    else
+        install ${PREFIX}/lib/libnfd.${dlext} ${libdir}/libnfd_gtk.${dlext}
+    fi
+    rm -r ${PREFIX}
+else
+    ln -s ${prefix}/lib/libnfd.${dlext} ${libdir}/libnfd_gtk.${dlext}
+fi
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = filter!(p -> arch(p) != "armv6l", supported_platforms())
+
+dbus_gtk = filter(p->Sys.islinux(p)||Sys.isfreebsd(p), platforms)
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libnfd", :libnfd)
+    LibraryProduct("libnfd_gtk", :libnfd_gtk)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    BuildDependency(PackageSpec(name="Xorg_xorgproto_jll", uuid="c4d99508-4286-5418-9131-c86396af500b"); platforms=platforms)
+    BuildDependency(PackageSpec(name="GTK3_jll", uuid="77ec8976-b24b-556a-a1bf-49a033a670a6"); platforms=platforms)
```

Comment:
I still don't understand why this isn't a regular Dependency.

Reply:
Given that for all platforms there is an alternative (native API for Windows/Apple, portal for Linux/FreeBSD), I decided to support the GTK3 backend only through a package extension.

I don't want the `NativeFileDialog.jl` (the user facing API of this artifact) to install unnecessary packages (strictly unnecessary for Windows/Apple, not necessary for most Linux/FreeBSD) for users when there is an alternative for basically everyone.

Reply:
Though, for the support of Portal in FreeBSD/Linux to be complete, there is still [this](https://github.com/JuliaPackaging/Yggdrasil/pull/8884) PR to update `Dbus` to be merged.

Reply:
Huh? I still don't understand. You are telling it that GTK3 is available, and it can link against it, so it will be. That should lead to a problem at runtime because the dependency is missing.

IMO, you should just provide these for Linux and FreeBSD unconditionally. The weight of these dependencies doesn't seem too large in my opinion.

Reply:
There are two artifacts generated here:

 - Portal/Native one, will use Dbus in Linux/FreeBSD and native API's on Windows/Apple, this is `libnfd`.
 - GTK3, will use `GTK3_jll` to call on a file dialog on all supported platform, this is `libnfd_gtk`.

I plant to use the first in the main package, as it covers _most_ of user needs (probably all if you do not count people who are installing a _really extremely terse_ `X` environment on Linux/FreeBSD).

I guess that importing `libnfd_gtk` will fail, because `dlopen`ing at `_init_` time will fail to link to GTK3. Is there a way to not open it at `_init_` time? I plan to program `NativeFileDialog.jl` use the GTK3 backend only as a package extension.

GTK3 is **NOT** necesary, not for basically _all users_. Installing and importing it cost time, disk and network, for no benefit. Having the GTK3 backend to work as an extension _does make this package lighter_, only incurring in _very light_ inconvenience when there is no other alternative for this to work.

I don't understand why this plan is not sound to you. It covers all possible users, benefiting most for an extremely small (`add GTK3`) cost against the very few that decided to not have `dbus` and `portal` installed in their systems (which should be _really, really, few people_).

I do not have my mind set in stone, by any means. I'm sorry if I sound stuborn. I'm just not yet capable of grasping if the solution I propose goes against some design of the registry, or if I'm being unable to transmit what I think is a reasonable compromise. 😅

Reply:
@giordano friendly poke :slightly_smiling_face:

Reply:
(p.s. @Suavesito-Olimpiada you could update this to 1.2.1 now)

Reply:
> Is there a way to not open it at `_init_` time?

Pass `dont_dlopen`: https://docs.binarybuilder.org/stable/reference/#BinaryBuilderBase.LibraryProduct

> I don't understand why this plan is not sound to you.

Because in my mind library `libA` in package `PackageA` links dynamically to `libB` in `PackageB` implies that `PackageA` depends on `PackageB`.  If this is only a platform-specific issue, we already allow for platform-specific [`Dependency`](https://docs.binarybuilder.org/stable/reference/#BinaryBuilderBase.Dependency)'s.